### ISSUE TITLE
[Feature]: Add EpiplexityScore

### ISSUE THREAD

**Issue by Mec-iS** [Mec-iS/arrowspace-rs#118]
### Problem Statement

The paper’s [**learnable novelty (epiplexity)**](https://arxiv.org/pdf/2607.18433) can become a principled new `IndexScorePhase` for ArrowSpace: not “how spectrally rough is this vector?” but “how much *stable, compressible new structure* does this item, batch, trajectory, or representation add for a bounded observer?” This is a strong conceptual fit with graph wiring, provided ArrowSpace treats the graph as part of the observer and does not equate high Laplacian energy with novelty.[1]

## Core correspondence

Graph wiring first converts a vector dataset into a feature-space graph and Laplacian; ArrowSpace currently uses Rayleigh/Dirichlet-style energy, bounded into a lambda-like index, to describe an item’s roughness relative to learned feature co-variation.  The new paper instead defines useful novelty as the **description length of the model a bounded learner can extract**, explicitly excluding both trivial regularity and irreducible noise; its practical estimator is a frozen random reservoir plus closed-form ridge readout, with a log-determinant score of the readout’s singular values.[1]

| Concept | Graph Wiring / ArrowSpace | Learnable Novelty paper | Combined interpretation |
|---|---|---|---|
| Object | Feature manifold induced from vectors | Input-output relation observed by bounded learner | A graph-conditioned data-generating relation |
| Signal | Dirichlet/Rayleigh roughness \(x^\top Lx / x^\top x\) | Compressible, independent structure in a learned readout | Structure that is both graph-relevant and bounded-observer learnable |
| Failure mode | A rare or high-energy point can be noise | Pure random noise earns little score | Require novelty to survive a bounded predictive/compression test |
| Useful output | Fast scalar for search, filtering, drift, and routing | Scalar objective for analysis, representation learning, and exploration | A capability-specific score attached to one shared wired graph |

The key synthesis is: **tau/energy measures position and roughness on the wired manifold; epiplexity measures whether an observer can turn variation associated with that position into reusable structure.**[1]

## A new score phase

Add a pluggable `LearnableNoveltyScore` (or `EpiplexityScore`) alongside `TauModeScore`, `FeatureSpectralScore`, `FiedlerVectorScore`, `BoundaryEnergyScore`, and `EnergyDiffusionScore`. This preserves the proposed “same pipeline, swappable capability” architecture while giving ArrowSpace an explicit bounded-complexity objective.[2]

```text
vectors
  -> ClusterPhase
  -> SamplingOrCompressionPhase
  -> GraphLaplacianPhase
  -> IndexScorePhase<T>

where T can be:
  TauModeScore              -> search / spectral prefiltering
  FeatureSpectralScore      -> analysis / feature circuits
  FiedlerVectorScore        -> partitioning / clustering
  BoundaryEnergyScore       -> active-learning classification
  EnergyDiffusionScore      -> graph-native compression
  LearnableNoveltyScore     -> curation, discovery, intrinsic reward
```

`LearnableNoveltyScore` should expose **two different products**:

- **Global epiplexity:** “How much learnable structure does this wired dataset/version contain?”
- **Marginal epiplexity:** “How much learnable structure does this item, cluster, batch, or candidate transformation contribute?”

The global score follows the paper’s reservoir-and-ridge construction:

\[
S_\phi(Y \mid X) =
\frac{1}{2}\log_2 \det(I + \eta W_\lambda W_\lambda^\top)
\]

where \(W_\lambda\) is a ridge-optimal readout from frozen reservoir features of \(X\) to targets \(Y\). 

For streaming ArrowSpace operations, the usable item-level value should be a **marginal contribution**, such as

\[
\Delta S_i = S_\phi(\mathcal{D}_{1:i}) - S_\phi(\mathcal{D}_{1:i-1}),
\]

or a stable approximation based on ridge leverage and a rank-one update. This respects the paper’s prequential logic: an item is valuable when it changes the learnable model rather than merely being distant, large, or noisy. 

## Make the graph the observer

The most distinctive ArrowSpace version is not epiplexity applied directly to raw embeddings. It is **graph-conditioned epiplexity**.

For each item vector \(x_i\), derive graph-native targets from the wired feature Laplacian \(L_F\), for example:

\[
Y_i =
\left[
x_i L_F,\;
x_i L_F^2,\;
\mathrm{Rayleigh}(x_i),\;
\mathrm{diffusion}_t(x_i)
\right].
\]

Then use a frozen reservoir to ask: *what aspects of a vector’s graph response can a bounded readout reliably recover from its embedding, local neighbourhood, metadata, or prior state?*

This creates a useful division of labour:

- `TauModeScore` identifies energy-band compatibility for fast candidate filtering.
- `LearnableNoveltyScore` estimates whether a candidate adds **structured novelty** to that band.
- `EnergyDiffusionScore` decides which redundant structures can be compressed away without losing unusual but coherent subcentroids. ArrowSpace’s existing energy pipeline already includes centroid diffusion/splitting and optical compression, so this is an adjacent rather than replacement capability.[2]

### Data curation

Use marginal epiplexity to choose training data that are neither duplicates nor incoherent outliers. Existing density-adaptive sampling already preserves sparse regions more readily than saturated dense regions; learnable novelty could refine this by retaining points that alter a bounded graph-conditioned model rather than all low-density points.[2]

A concrete policy is:

- Evict low-\(\lambda\), low-\(\Delta S\) items: common, graph-smooth, and observer-redundant.
- Review high-\(\lambda\), low-\(\Delta S\) items: graph-rough but potentially noise, corruption, or adversarial material.
- Retain high-\(\lambda\), high-\(\Delta S\) items: rare, structurally coherent cases that may improve coverage and OOD robustness.
- Retain moderate-\(\lambda\), high-\(\Delta S\) items: learnable bridges, emerging domains, and useful compositional examples.

This is more discriminating than novelty-by-distance and more operationally meaningful than an energy-only cache policy. The paper’s motivation is exactly that raw surprise can be maximized by noise, whereas learnable novelty excludes surprise a finite observer cannot absorb. 

### RAG and search

A graph-conditioned novelty score could provide a **second-stage diversity and support control** after tau-band filtering and cosine reranking. Graph Wiring already frames tau-aware retrieval as a way to prioritize candidates compatible with the feature manifold before more expensive semantic ranking.[1]

One candidate ranking function is:

\[
\mathrm{rank}(q,d)
=
\alpha\,\mathrm{cos}(q,d)
+
\beta\,\mathrm{tau\_compatibility}(q,d)
+
\gamma\,\Delta S(d \mid \mathcal{C}),
\]

where \(\mathcal{C}\) is the already-selected context set. Here \(\Delta S\) rewards a document only when it adds recoverable structure to the retrieved bundle—not simply lexical novelty or embedding distance.

This could support:

- **Evidence-set construction:** retrieve mutually complementary sources rather than ten graph-near paraphrases.
- **OOD abstention:** high graph energy with negligible learnable support is a reason to lower confidence or ask for verification.
- **Long-tail retention:** preserve rare but systematic documents during index compression.
- **Corpus-growth alerts:** flag batches whose new structure is learnable but not represented in the current retrieval graph.

### Active learning

`BoundaryEnergyScore` identifies uncertain or rough items relative to known class centroids; `LearnableNoveltyScore` can decide whether labelling them would likely teach the system something reusable. This yields a two-axis queue:

| Candidate state | Interpretation | Action |
|---|---|---|
| High boundary energy, high learnable novelty | Potential new coherent subclass or decision boundary | Prioritize expert label |
| High boundary energy, low learnable novelty | Noise, ambiguity, corrupted input, or unresolved representation | Audit or defer |
| Low boundary energy, high learnable novelty | Hidden factor within a known region | Add targeted labels or contrastive pairs |
| Low boundary energy, low learnable novelty | Redundant example | Deprioritize |

The paper reports that its score promoted unsupervised class separation in an MNIST encoder and worked as an intrinsic exploration bonus in reinforcement learning, making this active-learning use a natural intermediate application—not a claim that the result automatically transfers to all embedding corpora. 

### Representation learning

The paper’s representation-learning experiment maximizes learnable novelty of an encoder’s representation against a frozen reservoir and reports emergent digit-class separation without training labels.  ArrowSpace could test a graph-aware variant in which the encoder is trained to maximize epiplexity **while preserving local graph topology**.

A conservative objective would be:

\[
\max_\theta
\;
S_\phi\big(E_\theta(X) \mid X\big)
-
\mu\,\mathcal{L}_{\mathrm{topology}}
-
\nu\,\mathcal{L}_{\mathrm{instability}}.
\]

The topology term prevents the encoder from manufacturing arbitrary separability; the stability term enforces agreement under augmentation, subsampling, and reasonable graph-construction settings. This is important because ArrowSpace’s Laplacian interpretation only remains reliable when local neighbourhood structure and leading spectral properties are stable under its engineering approximations.[1]

### Dynamical systems

For sequence, sensor, simulator, biological, or agent-trajectory data, the paper’s most direct formulation becomes available: set \(X_t\) to a local state or graph state and \(Y_t\) to a future window. The paper uses the incremental rise in epiplexity as a dense intrinsic reward, reporting improvements over task-reward-only baselines in nine of ten studied environments. 

Graph Wiring adds an appealing extension: compute novelty not only in raw state space but in the state’s **feature-graph diffusion signature**. This could help distinguish:

- chaotic sensor noise: high entropy, low learnability;
- dead steady states: low entropy, low learnability;
- repeatable modes, transitions, and interacting motifs: high learnable novelty.

Potential initial domains include industrial anomaly detection, biological morphology trajectories, market-regime monitoring, simulator discovery, and autonomous scientific experimentation.

### Dataset fingerprints

ArrowSpace already positions spectra, graph topology, and index distributions as dataset fingerprints for compatibility and drift monitoring.  A global epiplexity statistic can enrich this fingerprint with a bounded-observer complexity coordinate:[3]

```text
fingerprint =
  truncated Laplacian spectrum
  + tau / Rayleigh distribution
  + cluster and diffusion statistics
  + epiplexity S_phi
  + marginal-epiplexity histogram
  + stability under reservoir seeds and subsampling
```

This gives a useful distinction between two datasets with similar spectra:

- Dataset A may be structurally rich in a stable, learnable way.
- Dataset B may have similar spectral roughness caused largely by noise, artifacts, or idiosyncratic variance.

## Implementation shape

A clean trait boundary would keep graph construction separate from scoring:

```rust
trait IndexScorePhase {
    type FitState;
    type ItemScore;
    type GlobalScore;

    fn fit(
        &self,
        data: &ArrowData,
        graph: &GraphLaplacian,
        topology: &TopologyState,
    ) -> Self::FitState;

    fn score_items(
        &self,
        state: &Self::FitState,
        data: &ArrowData,
        graph: &GraphLaplacian,
    ) -> Vec<Self::ItemScore>;

    fn score_global(&self, state: &Self::FitState) -> Self::GlobalScore;
}
```

For `LearnableNoveltyScore`, `FitState` should store:

- Reservoir seed and fixed random map specification.
- Feature/target normalisation constants.
- Ridge parameter \(\lambda\), spectral resolution \(\eta\), and reservoir width.
- Readout factorisation sufficient for efficient rank-one updates.
- Explicit definition of \(X\) and \(Y\), including any graph diffusion horizons.
- Held-out and stability diagnostics.

This suits ArrowSpace’s existing construction pattern: clustering, optional sampling, optional projection, Laplacian construction, then index computation. The current builder already separates sampling, dimensionality reduction, graph construction, tau synthesis, and energy-specific diffusion/compression paths, so a score-phase interface is a natural generalisation rather than a rewrite.[2]

## Critical safeguards

The paper’s score is **observer-relative**: changing reservoir width, random features, ridge penalty, target scale, or target construction changes what counts as learnable. It should therefore never be presented as an observer-free “intrinsic complexity” truth. 

I would make these non-negotiable:

- **Frozen observer:** do not train the reservoir per item or batch; otherwise the score can become a capacity meter rather than a comparable novelty measure.
- **Out-of-sample evaluation:** fit or update on one split/window and measure marginal predictive structure on another; otherwise self-generated graph targets can be trivially memorized.
- **Multi-seed confidence intervals:** report score stability over reservoir seeds, graph seeds, subsamples, and reasonable \(k\)/bandwidth settings.
- **Magnitude controls:** compare against norm, raw entropy, cosine distance, Rayleigh energy, and density; the paper itself used a state-magnitude control in RL to distinguish meaningful structure from scale chasing. 
- **No backpropagation through discrete rebuilding at first:** for representation training, hold cluster assignments and graph structure fixed for several steps, or use a differentiable soft graph approximation. Re-cluster periodically outside the differentiable loop.
- **Task-level validation:** accept the score only if it improves downstream retrieval diversity, curation efficiency, label efficiency, or OOD detection—not merely its own value.

## Best first experiment

The lowest-risk proof of value is **offline graph-conditioned curation**, not end-to-end representation optimization.

1. Start with a corpus that has downstream relevance labels or retrieval judgments.
2. Build the standard ArrowSpace graph once, including the current clustering/sampling and Laplacian procedure.
3. Define \(X\) as each item’s embedding plus local graph descriptors; define \(Y\) as its multi-scale graph-diffusion response and optionally metadata-derived targets.
4. Compute per-item approximate \(\Delta S\) with a frozen reservoir, then compare four retention policies: random, density-only, energy-only, and energy-plus-learnable-novelty.
5. Rebuild a fixed-budget index from each retained subset and evaluate retrieval quality, tail stability, duplicate rate, OOD behaviour, and score stability.

A successful result would be: **at equal storage and indexing budget, the hybrid policy preserves or improves retrieval and OOD performance while retaining fewer redundant items and fewer high-energy/noise-like artifacts.** That directly tests the proposed bridge between ArrowSpace’s structural graph signal and the paper’s distinction between useful novelty and unlearnable surprise.[1]

### Proposed Solution

Have an EpiplexityScore that sits besides TauMode for the applications mentioned above

### Priority

Medium - Would be helpful

### Estimated Complexity

Unknown - Not sure about implementation

### Feature Type

- [ ] API enhancement
- [ ] Performance improvement
- [x] New functionality
- [ ] Developer experience improvement
- [ ] Documentation enhancement
- [ ] Configuration option
- [ ] Integration with external service

### Code of Conduct

- [x] I agree to follow this project's Code of Conduct
- [x] I have searched existing issues to avoid duplicates
- [ ] I am willing to help implement this feature if guidance is provided
